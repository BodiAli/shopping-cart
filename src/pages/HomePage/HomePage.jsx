import { useLoaderData } from "react-router-dom";
import getData from "../../utils/fetchData";
import back from "../../assets/back.svg";
import forward from "../../assets/forward.svg";
import styles from "./HomePage.module.css";

function HomePage() {
  const data = useLoaderData();
  console.log(data);

  return (
    <main className={styles.main}>
      <div className={styles.headerContainer}>
        <h2>
          Welcome to GameVault
          <br />
          Your Ultimate Video Game Destination
        </h2>
        <p>
          Discover a vast selection of video games across all genres and platforms. Whether you&apos;re a
          casual gamer or a hardcore enthusiast, we have everything you need to elevate your gaming experience
        </p>
      </div>
      <div className={styles.carousel}>
        <button>
          <img src={back} alt="back" />
        </button>
        <div className={styles.card}>
          <div className={styles.imageContainer}>
            <div
              className={styles.backGround}
              style={{ backgroundImage: `url(${data[0].sample_cover.thumbnail_image})` }}
            ></div>
            <img src={data[0].sample_cover.thumbnail_image} alt="" />
          </div>
          <div className={styles.descriptionContainer}>
            <h3>{data[0].title}</h3>
            <div dangerouslySetInnerHTML={{ __html: data[0].description }}></div>
          </div>
        </div>
        <button>
          <img src={forward} alt="forward" />
        </button>
        <div className={styles.pagination}>
          <div className={styles.circle}></div>
          <div className={styles.circle}></div>
          <div className={styles.circle}></div>
          <div className={styles.circle}></div>
        </div>
      </div>
    </main>
  );
}

async function loader() {
  const data = await getData("/random?limit=5&format=normal&");
  if (!data) {
    throw new Error("Unable to fetch data");
  }
  return data.games;
}

HomePage.loader = loader;
export default HomePage;
