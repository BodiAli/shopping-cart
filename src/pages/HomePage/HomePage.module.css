.main {
  background-color: var(--green);
  padding: 4rem 0rem;
  display: flex;
  flex-direction: column;
  gap: 50px;
  min-height: 100vh;
}

.headerContainer {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 20px;
  text-align: center;
  padding: 2rem;
  background-color: var(--darkerGreen);
  box-shadow: inset 0px 0px 10px 5px var(--green);
}

.headerContainer > h2 {
  font-family: "Merriweather", Times, serif;
  font-size: 2rem;
}

.headerContainer > p {
  font-size: 1.3rem;
}

.carousel {
  display: grid;
  grid-template-columns: 50px 1fr 50px;
  grid-template-rows: 1fr 40px;
  padding: 0;
  box-shadow: 2.8px 2.8px 2.2px rgba(0, 0, 0, 0.02), 6.7px 6.7px 5.3px rgba(0, 0, 0, 0.028),
    12.5px 12.5px 10px rgba(0, 0, 0, 0.035), 22.3px 22.3px 17.9px rgba(0, 0, 0, 0.042),
    41.8px 41.8px 33.4px rgba(0, 0, 0, 0.05), 100px 100px 80px rgba(0, 0, 0, 0.07);
}

.carousel:hover {
  outline: 5px double var(--darkerGray);
}

.card {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  height: 100%;
  cursor: pointer;
}

.imageContainer {
  position: relative;
  width: 100%;
}

.imageContainer > .backGround {
  position: absolute;
  width: 100%;
  height: 100%;
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  z-index: 1;
  opacity: 0.5;
  filter: blur(2px);
}

.imageContainer > img {
  width: 100%;
  height: 300px;
  object-fit: contain;
  position: relative; /* Add stacking contest */
  z-index: 2;
  filter: drop-shadow(2.8px 2.8px 10px rgba(0, 0, 0, 0.775));
}

.descriptionContainer {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 10px;
  padding: 0.7rem;
  background-color: var(--gray);
  width: 100%;
  min-height: 258px;
}

.descriptionContainer > h3 {
  font-size: 2rem;
  font-family: "Merriweather", Times, serif;
}

.descriptionContainer > .price {
  font-size: 1.2rem;
}

.descriptionContainer > div {
  display: flex;
  flex-direction: column;
  gap: 7px;
  line-height: 1.3;
  max-height: 200px;
  overflow-y: auto;
  scrollbar-width: thin;
}

.shopContainer {
  background-color: #a7e4cea8;
  display: flex;
  padding: 3rem 2rem;
  justify-content: center;
  align-items: center;
}

.shopNowLink {
  text-decoration: none;
  color: black;
  background: linear-gradient(270deg, #00c896, #7e7c7c);
  background-size: 400% 400%;
  border: none;
  border-radius: 3px;
  padding: 1rem 2rem;
  font-family: "Merriweather", Times, serif;
  font-size: 1.8rem;
  cursor: pointer;
  animation: gradient-animation 18s ease infinite;
}

.shopNowLink:hover {
  filter: brightness(0.8);
}

.shopNowLink:active {
  filter: brightness(1.1);
}

@keyframes gradient-animation {
  0% {
    background-position: 0% 50%;
  }
  50% {
    background-position: 100% 50%;
  }
  100% {
    background-position: 0% 50%;
  }
}
