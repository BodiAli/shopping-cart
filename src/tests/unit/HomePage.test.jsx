import { render, screen } from "@testing-library/react";
import { createMemoryRouter, RouterProvider } from "react-router-dom";
import { vi } from "vitest";
import HomePage from "../../pages/HomePage/HomePage";

const games = {
  games: [
    {
      title: "Fifa game",
      sample_cover: {},
    },
    {
      title: "COD game",
      sample_cover: {},
    },
    {
      title: "Minecraft game",
      sample_cover: {},
    },
  ],
};

const MOCK_GAMES = games;

// eslint-disable-next-line no-undef
global.fetch = vi.fn(() =>
  Promise.resolve({
    ok: true,
    json: vi.fn(() => Promise.resolve(MOCK_GAMES)),
  })
);

const router = createMemoryRouter([
  {
    path: "/",
    element: <HomePage />,
    loader: HomePage.loader,
  },
]);

const wrapper = () => <RouterProvider router={router} />;

vi.mock("../../utils/generatePrice", () => {
  return {
    default: vi.fn(() => "50.00"),
  };
});

describe("Home page default component", () => {
  test("should render correct elements", () => {
    const { container } = render(<HomePage />, { wrapper });
    expect(container).toMatchSnapshot();
  });

  test("should render an intro header", () => {
    render(<HomePage />, { wrapper });
    expect(screen.getByRole("heading", { name: /welcome to gameVault/i })).toBeInTheDocument();
  });

  test("should render a description of the site", () => {
    render(<HomePage />, { wrapper });
    expect(screen.getByText(/discover a vast selection/i)).toBeInTheDocument();
  });

  test("should render games", () => {
    render(<HomePage />, { wrapper });
    const games = screen.getAllByRole("heading", { name: /game$/i });
    games.forEach((game) => {
      expect(game).toBeInTheDocument();
    });
  });

  test("each game should have a price", () => {
    render(<HomePage />, { wrapper });
    const prices = screen.getAllByText(/50.00/);

    prices.forEach((price) => {
      expect(price).toBeInTheDocument();
    });
  });

  test("should render an image carousel", () => {
    render(<HomePage />, { wrapper });
    const games = screen.getAllByRole("heading", { name: /game$/i });
    expect(screen.getByTestId("carousel")).toBeInTheDocument();
    games.forEach((game) => {
      expect(screen.getByTestId("carousel")).toContainElement(game);
    });
  });

  test("should render a nav that links to shop", () => {
    render(<HomePage />, { wrapper });
    const shopNowLink = screen.getByRole("link", { name: /shop now/i });
    expect(shopNowLink).toBeInTheDocument();
    expect(shopNowLink).toHaveAttribute("href", "/shop/games");
  });
});
