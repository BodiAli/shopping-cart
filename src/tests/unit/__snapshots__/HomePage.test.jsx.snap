// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home page default component > should render correct elements 1`] = `
<div>
  <main
    class="_main_ea6e48"
  >
    <div
      class="_headerContainer_ea6e48"
    >
      <h2>
        Welcome to GameVault
        <br />
        Your Ultimate Video Game Destination
      </h2>
      <p>
        Discover a vast selection of video games across all genres and platforms. Whether you're a casual gamer or a hardcore enthusiast, we have everything you need to elevate your gaming experience
      </p>
    </div>
    <section
      aria-roledescription="carousel"
      class="splide _carousel_ea6e48 splide--loop splide--ltr splide--draggable is-active is-initialized"
      data-testid="carousel"
      id="splide01"
    >
      <div
        class="splide__arrows splide__arrows--ltr"
      >
        <button
          aria-controls="splide01-track"
          aria-label="Go to last slide"
          class="splide__arrow splide__arrow--prev"
          type="button"
        >
          <svg
            focusable="false"
            height="40"
            viewBox="0 0 40 40"
            width="40"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m15.5 0.932-4.3 4.38 14.5 14.6-14.5 14.5 4.3 4.4 14.6-14.6 4.4-4.3-4.4-4.4-14.6-14.6z"
            />
          </svg>
        </button>
        <button
          aria-controls="splide01-track"
          aria-label="Next slide"
          class="splide__arrow splide__arrow--next"
          type="button"
        >
          <svg
            focusable="false"
            height="40"
            viewBox="0 0 40 40"
            width="40"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m15.5 0.932-4.3 4.38 14.5 14.6-14.5 14.5 4.3 4.4 14.6-14.6 4.4-4.3-4.4-4.4-14.6-14.6z"
            />
          </svg>
        </button>
      </div>
      <div
        aria-atomic="true"
        aria-live="off"
        class="splide__track splide__track--loop splide__track--ltr splide__track--draggable"
        id="splide01-track"
        style="padding-left: 0px; padding-right: 0px;"
      >
        <ul
          class="splide__list"
          id="splide01-list"
          role="presentation"
          style="transform: translateX(0px);"
        >
          <li
            aria-label="2 of 3"
            aria-roledescription="slide"
            class="splide__slide splide__slide--clone is-visible"
            id="splide01-clone01"
            role="tabpanel"
          >
            <div
              class="_card_ea6e48"
            >
              <div
                class="_imageContainer_ea6e48"
              >
                <div
                  class="_backGround_ea6e48"
                  style="background-image: url(undefined);"
                />
                <img
                  alt="COD game game poster"
                />
              </div>
              <div
                class="_descriptionContainer_ea6e48"
              >
                <h3>
                  COD game
                </h3>
                <p
                  class="_price_ea6e48"
                >
                  Price: 50.00$
                </p>
                <div />
              </div>
            </div>
          </li>
          <li
            aria-label="3 of 3"
            aria-roledescription="slide"
            class="splide__slide splide__slide--clone is-visible is-prev"
            id="splide01-clone02"
            role="tabpanel"
          >
            <div
              class="_card_ea6e48"
            >
              <div
                class="_imageContainer_ea6e48"
              >
                <div
                  class="_backGround_ea6e48"
                  style="background-image: url(undefined);"
                />
                <img
                  alt="Minecraft game game poster"
                />
              </div>
              <div
                class="_descriptionContainer_ea6e48"
              >
                <h3>
                  Minecraft game
                </h3>
                <p
                  class="_price_ea6e48"
                >
                  Price: 50.00$
                </p>
                <div />
              </div>
            </div>
          </li>
          <li
            aria-label="1 of 3"
            aria-roledescription="slide"
            class="splide__slide is-active is-visible"
            id="splide01-slide01"
            role="tabpanel"
          >
            <div
              class="_card_ea6e48"
            >
              <div
                class="_imageContainer_ea6e48"
              >
                <div
                  class="_backGround_ea6e48"
                  style="background-image: url(undefined);"
                />
                <img
                  alt="Fifa game game poster"
                />
              </div>
              <div
                class="_descriptionContainer_ea6e48"
              >
                <h3>
                  Fifa game
                </h3>
                <p
                  class="_price_ea6e48"
                >
                  Price: 50.00$
                </p>
                <div />
              </div>
            </div>
          </li>
          <li
            aria-label="2 of 3"
            aria-roledescription="slide"
            class="splide__slide is-visible is-next"
            id="splide01-slide02"
            role="tabpanel"
          >
            <div
              class="_card_ea6e48"
            >
              <div
                class="_imageContainer_ea6e48"
              >
                <div
                  class="_backGround_ea6e48"
                  style="background-image: url(undefined);"
                />
                <img
                  alt="COD game game poster"
                />
              </div>
              <div
                class="_descriptionContainer_ea6e48"
              >
                <h3>
                  COD game
                </h3>
                <p
                  class="_price_ea6e48"
                >
                  Price: 50.00$
                </p>
                <div />
              </div>
            </div>
          </li>
          <li
            aria-label="3 of 3"
            aria-roledescription="slide"
            class="splide__slide is-visible"
            id="splide01-slide03"
            role="tabpanel"
          >
            <div
              class="_card_ea6e48"
            >
              <div
                class="_imageContainer_ea6e48"
              >
                <div
                  class="_backGround_ea6e48"
                  style="background-image: url(undefined);"
                />
                <img
                  alt="Minecraft game game poster"
                />
              </div>
              <div
                class="_descriptionContainer_ea6e48"
              >
                <h3>
                  Minecraft game
                </h3>
                <p
                  class="_price_ea6e48"
                >
                  Price: 50.00$
                </p>
                <div />
              </div>
            </div>
          </li>
          <li
            aria-label="1 of 3"
            aria-roledescription="slide"
            class="splide__slide splide__slide--clone is-active is-visible"
            id="splide01-clone03"
            role="tabpanel"
          >
            <div
              class="_card_ea6e48"
            >
              <div
                class="_imageContainer_ea6e48"
              >
                <div
                  class="_backGround_ea6e48"
                  style="background-image: url(undefined);"
                />
                <img
                  alt="Fifa game game poster"
                />
              </div>
              <div
                class="_descriptionContainer_ea6e48"
              >
                <h3>
                  Fifa game
                </h3>
                <p
                  class="_price_ea6e48"
                >
                  Price: 50.00$
                </p>
                <div />
              </div>
            </div>
          </li>
          <li
            aria-label="2 of 3"
            aria-roledescription="slide"
            class="splide__slide splide__slide--clone is-visible"
            id="splide01-clone04"
            role="tabpanel"
          >
            <div
              class="_card_ea6e48"
            >
              <div
                class="_imageContainer_ea6e48"
              >
                <div
                  class="_backGround_ea6e48"
                  style="background-image: url(undefined);"
                />
                <img
                  alt="COD game game poster"
                />
              </div>
              <div
                class="_descriptionContainer_ea6e48"
              >
                <h3>
                  COD game
                </h3>
                <p
                  class="_price_ea6e48"
                >
                  Price: 50.00$
                </p>
                <div />
              </div>
            </div>
          </li>
        </ul>
      </div>
      <ul
        aria-label="Select a slide to show"
        class="splide__pagination splide__pagination--ltr"
        role="tablist"
      >
        <li
          role="presentation"
        >
          <button
            aria-controls="splide01-slide01"
            aria-label="Go to slide 1"
            aria-selected="true"
            class="splide__pagination__page is-active"
            role="tab"
            type="button"
          />
        </li>
        <li
          role="presentation"
        >
          <button
            aria-controls="splide01-slide02"
            aria-label="Go to slide 2"
            class="splide__pagination__page"
            role="tab"
            tabindex="-1"
            type="button"
          />
        </li>
        <li
          role="presentation"
        >
          <button
            aria-controls="splide01-slide03"
            aria-label="Go to slide 3"
            class="splide__pagination__page"
            role="tab"
            tabindex="-1"
            type="button"
          />
        </li>
      </ul>
    </section>
    <section
      class="_shopContainer_ea6e48"
    >
      <a
        class="_shopNowLink_ea6e48"
        href="/shop/games"
      >
        Shop Now!
      </a>
    </section>
  </main>
</div>
`;
